src/openlcb/MemoryConfigClient.hxx

```diff
@@ -270,7 +296,156 @@ private:
         cleanup_read();
         return return_ok();
     }
-    
+
+    Action do_write()
+    {
+        dgClient_ = full_allocation_result(dg_service()->client_allocator());
+        offset_ = request()->address;
+        payloadOffset_ = 0;
+        memoryConfigHandler_->set_client(&responseFlow_);
+        return call_immediately(STATE(send_next_write));
+    }
+
+    Action send_next_write()
+    {
+        return allocate_and_call(
+            dg_service()->iface()->dispatcher(), STATE(send_write_datagram));
+    }
+
+    Action send_write_datagram()
+    {
+        auto *b = get_allocation_result(dg_service()->iface()->dispatcher());
+        b->set_done(bn_.reset(this));
+        unsigned sz = request()->payload.size() - payloadOffset_;
+        if (sz > 64)
+            sz = 64;
+        writeLength_ = sz;
+        b->data()->reset(Defs::MTI_DATAGRAM, node_->node_id(), request()->dst,
+            MemoryConfigDefs::write_datagram(request()->memory_space, offset_,
+                             request()->payload.substr(payloadOffset_, sz)));
+        isWaitingForTimer_ = 0;
+        responseCode_ = DatagramClient::OPERATION_PENDING;
+        dgClient_->write_datagram(b);
+        return wait_and_call(STATE(write_complete));
+    }
+
+    Action write_complete()
+    {
+        if (!(dgClient_->result() & DatagramClient::OPERATION_SUCCESS))
+        {
+            // some error occurred.
+            return handle_write_error(dgClient_->result());
+        }
+        if (responseCode_ & DatagramClient::OPERATION_PENDING)
+        {
+            isWaitingForTimer_ = 1;
+            return sleep_and_call(
+                &timer_, SEC_TO_NSEC(3), STATE(write_response_timeout));
+        }
+        else
+        {
+            return call_immediately(STATE(write_response_timeout));
+        }
+    }
+
+    Action write_response_timeout()
```

Comment:
There seems to be a lot of code repeated here from read_response_timeout(). It would be good to refactor to have methods that handle looking at the bytes. For example, to pull out the address. DRY!

Reply:
I refactored some of the repeated code into helper functions, but it did not get all that much simpler.